### ISSUE TITLE
chore: upgrade Ray from 2.51.1 to 2.55.1 for GPU efficiency and scheduling improvements

### ISSUE THREAD

**Issue by sallycr** [michelangelo-ai/michelangelo#1272]
**Is your feature request related to a problem? Please describe.**
The project currently pins Ray at `2.51.1` (locked in `poetry.lock`) with a constraint of `^2.41.0` in `pyproject.toml`. Ray `2.55.1` (released Apr 22, 2026) is the latest stable release and includes significant improvements to GPU resource efficiency and task scheduling that are directly relevant to Michelangelo's ML training and inference workloads.

**Describe the solution you'd like**

Upgrade `poetry.lock` to resolve Ray to `2.55.1` and validate against the integration test suite.

```toml
# python/pyproject.toml — no change needed (^2.41.0 already allows 2.55.x)
ray = {extras = ["default"], version = "^2.41.0", optional = true}
```

```bash
cd python && poetry update ray
```

**Key improvements from 2.51.1 → 2.55.1**

### GPU Efficiency (2.54–2.55)
- **Don't reserve GPU budget for non-GPU tasks** (#59789) — prevents GPU memory being held for CPU-only tasks; directly improves cluster utilization on mixed workloads.
- **GPU stage autoscaling** (#61130) — automatic scaling of GPU-intensive pipeline stages based on real demand, reducing idle GPU time.
- **GPU shuffle support via `rapidsmpf`** (#61371) — GPU-accelerated distributed data shuffling for large dataset preprocessing.
- **Fix GPU autoscaling when `max_actors` is set** (#59632) — unblocks GPU autoscaling on actor pools with a configured ceiling.

### Scheduling (2.52–2.55)
- **Include logical memory in scheduling decisions** (#60774) — prevents task over-subscription by factoring logical memory into placement.
- **Replica routing data structure optimizations** (#60139) — O(1) pending-request lookup, cached replica lists, lazy cleanup; reduces scheduling overhead at high QPS.
- **Revise resource allocator scheduling to factor in pending task outputs** (#60639) — improves look-ahead for downstream tasks.
- **Queue-based autoscaling policy** (#59548, #60851) — scales based on both HTTP in-flight load and queued tasks; better burst handling.
- **Heap-based actor ranking + actor pool map improvements** (#61996, #62114) — faster actor selection on large pools.
- **Sequence in-order actor tasks per concurrency group** (#61082) — reduces global lock contention, improves throughput for concurrent workloads.
- **Utilization-based cluster autoscaler** for Ray Data (`RAY_DATA_CLUSTER_AUTOSCALER=V2`, #59353) — smarter scale-up/down decisions tied to actual utilization.

### Data / IO Performance (2.52–2.55)
- **Zero-copy map operations by default** (#58285) — eliminates unnecessary data copies across all map transforms.
- **PyTorch zero-copy tensor serialization** via `RAY_ENABLE_ZERO_COPY_TORCH_TENSORS` (#57639) — reduces tensor transfer overhead between workers.
- **Parallelize `DefaultCollateFn` and `arrow_batch_to_tensors`** (#58821) — faster batch collation for training loops.
- **Streaming repartition + HashShuffle optimizations** (#58728, #58603) — lower overhead on large dataset repartitioning.
- **Optimize concat tables and PyArrow schema hashing** (#61315, #62108) — faster batch concatenation.

**Describe alternatives you've considered**
- Staying on `2.51.1`: misses the GPU budget fix (#59789) and the scheduling over-subscription fix (#60774) which are directly impactful for training jobs that mix GPU and CPU tasks.
- Upgrading to `2.53.x` or `2.54.x`: partial improvement; `2.55.1` is the latest stable patch and includes the most complete set of fixes.

**Additional context**
- `pyproject.toml` already allows `^2.41.0`, so only `poetry.lock` needs updating — no version constraint change required.
- Ray changelog reference: https://github.com/ray-project/ray/releases
- Current locked version: `ray==2.51.1` (see `python/poetry.lock`)
- Target version: `ray==2.55.1` (PyPI, released Apr 22, 2026)

**Comment by krishpatel9** [comment]
For some more context on this, the primary improvement that related to GPU inference is in to Data/IO performance. With the earlier version of Ray, when we used very large/wide data the deserialization cost to load data was scaling with the batch size, so inference was bottlenecked.

**Comment by sallycr** [comment]
**Partial upgrade landed: `2.51.1` → `2.53.0`**

Ray `2.54.0` dropped Python 3.9 support (`requires Python >=3.10`). The project currently uses `python = "^3.9.0"` and the examples Dockerfile builds Python 3.9.18 from source, so `2.55.1` cannot be resolved without a Python version bump.

**What this PR delivers (2.52–2.53 improvements):**
- Zero-copy map operations by default
- PyTorch zero-copy tensor serialization (`RAY_ENABLE_ZERO_COPY_TORCH_TENSORS`)
- Utilization-based cluster autoscaler for Ray Data (`RAY_DATA_CLUSTER_AUTOSCALER=V2`)
- Streaming repartition + HashShuffle performance improvements
- Parallelized `DefaultCollateFn` / `arrow_batch_to_tensors`
- Concurrency lock removal for better throughput
- Ray Direct Transport NIXL integration (faster inter-node comms)

**To reach `2.55.1` (GPU stage autoscaling, GPU budget fix, O(1) replica routing):**
1. Upgrade `python/pyproject.toml`: `python = "^3.9.0"` → `python = ">=3.10,<4.0"`
2. Update `python/examples/Dockerfile`: `ENV PYTHON_VERSION=3.9.18` → `3.10.x`
3. Run `poetry lock` and validate

This should be a separate issue/PR as it is a breaking change for any users running Python 3.9.

**Comment by sallycr** [comment]
**Update:** revised approach — both versions now coexist via Python environment markers in `pyproject.toml`:

```toml
ray = [
  {extras = ["default"], version = "^2.53.0", python = "~3.9",   optional = true},
  {extras = ["default"], version = "^2.55.1", python = ">=3.10", optional = true},
]
```

Poetry resolves the correct variant at install time — Python 3.9 environments get `2.53.0`, Python 3.10+ environments get `2.55.1`. Both versions are locked in `poetry.lock`. No code changes required.
